### PR TITLE
KPM system issue #70

### PR DESCRIPTION
**Pull request by wwjbatista**

### CODE REVIEW

File: kpm/lib/kpm/formatter.rb [hunk 1]
```diff
@@ -62,24 +62,27 @@ def format_sha(sha)
...
 
-      # What we want to output
-      labels = [{:label => :plugin_name},
-                {:label => :plugin_key},
-                {:label => :type},
-                {:label => :group_id},
-                {:label => :artifact_id},
-                {:label => :packaging},
-                {:label => :versions, :formatter => VersionFormatter.name}]
+      if labels == nil
```
Comment (sbrossie) [inline kpm/lib/kpm/formatter.rb]: Nit: You could use `labels.nil?` (like you did at other places)

File: kpm/lib/kpm/system.rb [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    def show_plugin_information(destination)
+      inspector = KPM::Inspector.new
+      all_plugins = inspector.inspect(destination)
+
+      @formatter.format(all_plugins)
+    end
+
+    def get_kaui_version
+      gemfile = Dir['**' + File::SEPARATOR + 'kaui' + File::SEPARATOR + 'WEB-INF' + File::SEPARATOR + 'Gemfile']
```
Comment (sbrossie) [inline kpm/lib/kpm/system.rb]: I am wondering if the command should take an option to specify where this has been installed?

File: kpm/lib/kpm/system.rb [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        end
+
+      end
+
+      version
+    end
+
+    def get_killbill_version
+      file =  Dir['**' + File::SEPARATOR + 'ROOT' + File::SEPARATOR + 'META-INF' +  File::SEPARATOR + '**' + File::SEPARATOR + 'pom.properties']
+      version = nil
```
Comment (sbrossie) [inline kpm/lib/kpm/system.rb]: Same remark as KAUI: I am wondering if the command should take an option to specify where this has been installed?

@pierre Any thoughts?

File: kpm/lib/kpm/system.rb [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+      @formatter.format(environment,labels)
+
+      show_environment_information
+      show_os_information
+      show_java_system_information
+      show_plugin_information(destination)
+    end
+
+    def show_environment_information
+      java_version = `java -version 2>&1`.split("\n")[0].split('"')[1]
```
Comment (sbrossie) [inline kpm/lib/kpm/system.rb]: Might make sense to harden the code (e.g if java is not installed you will get an NPE, in which case we still want to output and for java we would 'no version found').

File: kpm/lib/kpm/formatter.rb [hunk 5]
```diff
@@ -62,24 +62,27 @@ def format_sha(sha)
     end
 
 
-    def format(all_plugins)
-      if all_plugins.size == 0
+    def format(data, labels = nil)
+      if data.size == 0
```
Comment (sbrossie) [inline kpm/lib/kpm/formatter.rb]: `data.nil? || data.size == 0` might be better?

File: kpm/lib/kpm/system.rb [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+                {:label => :value}]
+
+      @formatter.format(os,labels)
+    end
+
+    def show_java_system_information
+      java_system = Hash.new
+      property_count = 0;
+      `java -XshowSettings:properties -version 2>&1`.split("\n").each do |prop|
+        property_count += 1
```
Comment (sbrossie) [inline kpm/lib/kpm/system.rb]: Same remark on hardening here

File: kpm/lib/kpm/formatter.rb (hunk 5 above)
Comment (wwjbatista) [inline kpm/lib/kpm/formatter.rb]: Yes, I agree. Missed that one.

File: kpm/lib/kpm/formatter.rb (hunk 1 above)
Comment (wwjbatista) [inline kpm/lib/kpm/formatter.rb]: Yes, I agree.

File: kpm/lib/kpm/system.rb (hunk 2 above)
Comment (wwjbatista) [inline kpm/lib/kpm/system.rb]: I will put an option to permit the user to add a location and if the location is not entered then the directory search will look for the file and if the file do not exist then a message that say "Kaui not found" . What do you think?

File: kpm/lib/kpm/system.rb (hunk 3 above)
Comment (wwjbatista) [inline kpm/lib/kpm/system.rb]: Same as KAUI, but with the message "Killbill API not found"

File: kpm/lib/kpm/system.rb (hunk 6 above)
Comment (wwjbatista) [inline kpm/lib/kpm/system.rb]: Ok

File: kpm/lib/kpm/system.rb (hunk 4 above)
Comment (wwjbatista) [inline kpm/lib/kpm/system.rb]: Ok

File: kpm/lib/kpm/system.rb (hunk 3 above)
Comment (pierre) [inline kpm/lib/kpm/system.rb]: Agreed on having an option for the user to specify the Kill Bill and Kaui installation paths.

I would also suggest adding yet another option to specify the kpm.yml file used for the installation (see https://github.com/killbill/killbill-cloud/blob/master/kpm/install_example.yml for an example), as it should contain the right paths already.

**Comment by pierre** [comment]
@wwjbatista FYI, I'm getting the same error (`uninitialized constant KPM::System`).

Also, testing the script in a Docker container, I'm getting:

```
_____________________
| OS DETAIL | VALUE |
_____________________
|   /bin/sh |     1 |
_____________________
```

Looking pretty good otherwise!

@sbrossie Question regarding improvements:

* Should we add an option to turn the output into a machine readable format like JSON (current output is nice, but we might want to be able to script it out)?
* About Java properties, should we also print out the ones used at runtime (they would be extracted from the kpm.yml if present, and/or we could detect if a Java process is running)?
* Should we extract simple information from the machine? I'm thinking:
  * CPU info
  * Memory info
  * Disk space
  * Entropy available

File: kpm/lib/kpm/tasks.rb [hunk 7]
```diff
@@ -403,6 +403,16 @@ def inspect
           #puts all_plugins.to_json
           inspector.format(all_plugins)
         end
+        
+        method_option :destination,
+                      :type    => :string,
+                      :default => nil,
```
Comment (pierre) [inline kpm/lib/kpm/tasks.rb]: We should set a default value (for instance, I needed to add `--destination /var/lib/killbill/bundles` to make it work in our Docker image). I'm thinking either `/var/lib/killbill/bundles` (default in Docker) or `/var/tmp/bundles` (Kill Bill default). @sbrossie thoughts? Or should we search both?

**Comment by sbrossie** [comment]
@wwjbatista Code looks good, will merge PR. I will run some additional testing once this has been merged to confirm it works with at least our current use case (and if i see any issues, i'll report back and we can open new PR).

**Pull request merged**